Mathlib/Algebra/Operad/Perm.lean

```diff
@@ -0,0 +1,252 @@
+-/
+open Equiv
+
+/-- Extend a permutation on Fin n to a permutation on Fim (m+n+k), by acting as the identity
+ on the first m and last k elements. -/
+def PermFinPadLeftRight {n : ℕ} (p : Perm (Fin n)) (m k : ℕ) : (Perm (Fin (m + n + k))) :=
+  Perm.extendDomain p (p := fun x ↦ m ≤ x ∧ x < m + n) ⟨
+    fun x ↦ ⟨(Fin.natAdd m x).castAdd k, by simp [Fin.natAdd, Fin.addNat]⟩,
+    fun x ↦ ⟨(Fin.castLT x.1 (show x < n + m by omega)).subNat m (by simp [x.2.1]),
+      Nat.sub_lt_left_of_lt_add x.2.1 x.2.2⟩,
+    fun x ↦ by simp,
+    fun x ↦ by
+      ext
+      simp only [Fin.coe_subNat, Fin.coe_castLT, Fin.natAdd_mk, Fin.castAdd_mk]
+      omega
+    ⟩
+
+/-- PermfindPadTo is morally equivalent (`HEq`) to PermfinPadLeftRight, but with a different type.
+ It starts the permutation Sm at location k out of n, and creates a perm of length (n+m-1). -/
+@[irreducible]
+def PermFinPadTo {m : ℕ} (p : Perm (Fin m)) (n : ℕ) (k : Fin n) : Perm (Fin (n+m-1)) :=
+  have h : (k + m + (n - (k + 1))) = n + m - 1 := by
+    omega
+  h ▸ PermFinPadLeftRight p k (n-(k+1))
+
+section PermFinPadTo
+
+theorem PermFinPadTo_eq_PermFinPadLeftRight {m : ℕ} (p : Perm (Fin m)) (n : ℕ) (k) (x) {pf} :
+    (PermFinPadTo p n k ⟨x, pf⟩ : ℕ) = PermFinPadLeftRight p k (n-(k+1)) ⟨x, by omega⟩ := by
+  rw [PermFinPadTo]
+  have : n + m - 1 = ↑k + m + (n - (↑k + 1)) := by omega
+  congr! 1
+  congr!
+  simp
+
+/-- These three theorems specify the action of PermFinPadTo, based on whether the input is below,
+  within, or above the interval [m,m+n) that the permutation is mapped to -/
+theorem PermFinPadTo_eq_position {m : ℕ} (p : Perm (Fin m)) (n : ℕ) (k : Fin n) (x : Fin m) {pf} :
+    PermFinPadTo p n k ⟨k + x, pf⟩ = (k : ℕ) + (p x) := by
+  rw [PermFinPadTo_eq_PermFinPadLeftRight, PermFinPadLeftRight,
+    Equiv.Perm.extendDomain_apply_subtype]
+  · simp
+  · dsimp
+    constructor <;> omega
+
+theorem PermFinPadTo_lt_position {m : ℕ} (p : Perm (Fin m)) (n : ℕ) (k : Fin n) (x : ℕ) (h : x < k)
+    {pf} : PermFinPadTo p n k ⟨x, pf⟩ = x := by
+  rw [PermFinPadTo_eq_PermFinPadLeftRight, PermFinPadLeftRight,
+    Equiv.Perm.extendDomain_apply_not_subtype]
+  simp only [not_and, not_lt]
+  omega
+
+theorem PermFinPadTo_gt_position {m : ℕ} (p : Perm (Fin m)) (n : ℕ) (k : Fin n) (x : ℕ)
+    (h : x + 1 > k + m) {pf} : PermFinPadTo p n k ⟨x, pf⟩ = x := by
+  rw [PermFinPadTo_eq_PermFinPadLeftRight, PermFinPadLeftRight,
+    Equiv.Perm.extendDomain_apply_not_subtype]
+  simp only [not_and, not_lt]
+  omega
+
+end PermFinPadTo
+
+def PermFinPadAt_core {m n : ℕ} (p : Perm (Fin m)) (hn : 0 < n) (k : Fin m) (x : Fin (m+n-1)) :
+    Fin (m+n-1) :=
+  if h : k.1 ≤ x.1 ∧ x.1 ≤ k + n - 1 then
+    ⟨p k + x - k.1, by omega⟩
+  else (
+    let x' := if h₂ : x < k.1 then p ⟨x.1, h₂.trans k.2⟩ else p ⟨x.1 - (n-1), by omega⟩;
+    if h₃ : x' < p k then ⟨x', by omega⟩ else ⟨x'+n-1, by omega⟩
+  )
+
+theorem PermFinPadAt_core.LeftInverse {m n : ℕ} (p : Perm (Fin m)) (hn : 0 < n) (k : Fin m) :
```

Comment:
This is a very ugly long proof and I would very much like to see it cleaned up. I will try, later, I guess.